File: OmniCore.Py/nonce.cs
using OmniCore.py;

namespace OmniCore.Py
{

	public class Nonce
    {
        public static uint FAKE_NONCE = 0xD012FA62;
        private int lot;
        private int tid;
        private uint? lastNonce;
        private int seed;
        private int? ptr;
        private int nonce_runs;
        private uint[] table;

		public Nonce(int lot, int tid, uint? seekNonce = null, int seed = 0)
        {
            this.lot = lot;
            this.tid = tid;
            this.lastNonce = null;
            this.seed = seed;
            this.ptr = null;
            this.nonce_runs = 0;
            this._initialize();
			if (seekNonce != null)
            {
                while (this.lastNonce != seekNonce.Value)
                {
                    getNext(true);
                }
            }
        }

		public uint getNext(bool seeking = false)
        {
            if (!seeking && this.nonce_runs > 25)
            {
                this.lastNonce = FAKE_NONCE;
                return FAKE_NONCE;
            }
            var nonce = this.table[this.ptr.Value];
            this.table[this.ptr.Value] = this._generate();
            this.ptr = (int)(nonce & 0xF) + 2;
            this.lastNonce = nonce;
            this.nonce_runs += 1;
            return nonce;
        }

        public void reset()
        {
            this.nonce_runs = 255;
        }
        

	    public void sync(uint syncWord, int msgSequence)
        {
            var w_sum = (this.lastNonce & 0xFFFF) + (CrcUtil.crc16_table[msgSequence] & 0xFFFF)
                        + (this.lot & 0xFFFF) + (this.tid & 0xFFFF);
            this.seed = (int)((w_sum & 0xFFFF) ^ syncWord) & 0xff;
            this.lastNonce = null;
            this.nonce_runs = 0;
            this._initialize();
        }

        public uint _generate()
        {
            this.table[0] = ((this.table[0] >> 16) + (this.table[0] & 0xFFFF) * 0x5D7F) & 0xFFFFFFFF;
            this.table[1] = ((this.table[1] >> 16) + (this.table[1] & 0xFFFF) * 0x8CA0) & 0xFFFFFFFF;
            return (this.table[1] + (this.table[0] << 16)) & 0xFFFFFFFF;
        }

        public void _initialize()
        {
            this.table = new uint[18];
            this.table[0] = (((uint)this.lot & 0xFFFF) + 0x55543DC3 + ((uint)this.lot >> 16) + ((uint)this.seed & 0xFF)) & 0xFFFFFFFF;
            this.table[1] = (((uint)this.tid & 0xFFFF) + 0xAAAAE44E + ((uint)this.tid >> 16) + ((uint)this.seed >> 8)) & 0xFFFFFFFF;

            for (int i = 2; i < 18; i++)
                this.table[i] = this._generate();

            this.ptr = (int)((this.table[0] + this.table[1]) & 0xF) + 2;
            this.lastNonce = null;
        }
    }
}

File: OmniCore.Py/pyref/nonce.py
from .crc import crc16_table

FAKE_NONCE = 0xD012FA62


class Nonce:
    def __init__(self, lot, tid, seekNonce = None, seed = 0):
        self.lot = lot
        self.tid = tid
        self.lastNonce = None
        self.seed = seed
        self.ptr = None
        self.nonce_runs = 0
        self._initialize()
        if seekNonce is not None:
            while self.lastNonce != seekNonce:
                self.getNext(True)

    def getNext(self, seeking = False):
        if not seeking and self.nonce_runs > 200:
            self.lastNonce = FAKE_NONCE
            return FAKE_NONCE
        nonce = self.table[self.ptr]
        self.table[self.ptr] = self._generate()
        self.ptr = (nonce & 0xF) + 2
        self.lastNonce = nonce
        self.nonce_runs += 1
        return nonce

    def reset(self):
        self.nonce_runs = 255

    def sync(self, syncWord, msgSequence):
        w_sum = (self.lastNonce & 0xFFFF) + (crc16_table[msgSequence] & 0xFFFF) \
              + (self.lot & 0xFFFF) + (self.tid & 0xFFFF)
        self.seed = ((w_sum & 0xFFFF) ^ syncWord) & 0xff
        self.lastNonce = None
        self.nonce_runs = 0
        self._initialize()

    def _generate(self):
        self.table[0] = ((self.table[0] >> 16) + (self.table[0] & 0xFFFF) * 0x5D7F) & 0xFFFFFFFF
        self.table[1] = ((self.table[1] >> 16) + (self.table[1] & 0xFFFF) * 0x8CA0) & 0xFFFFFFFF
        return (self.table[1] + (self.table[0] << 16)) & 0xFFFFFFFF

    def _initialize(self):
        self.table = [0]*18
        self.table[0] = ((self.lot & 0xFFFF) + 0x55543DC3 + (self.lot >> 16) + (self.seed & 0xFF)) & 0xFFFFFFFF
        self.table[1] = ((self.tid & 0xFFFF) + 0xAAAAE44E + (self.tid >> 16) + (self.seed >> 8)) & 0xFFFFFFFF

        for i in range(2, 18):
            self.table[i] = self._generate()

        self.ptr = ((self.table[0] + self.table[1]) & 0xF) + 2
        self.lastNonce = None
